[WebProfilerBundle] [WIP] Add help in Web Profiler

maidmaid opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Hello,

Symfony is a truly ideal framework to teach young future web developer to create a complete and robust web application, thanks to the excellent quality of its documentation.

Several links of this documentation are already present directly in the source code, making learning and understanding of Symfony easier. In the same idea, this PR proposes to add links of documentation in the Web Profiler, to allow a better pedagogical experiment.

The next first example adds cookie doc link:

![screenshot from 2016-12-24 22-21-10](https://cloud.githubusercontent.com/assets/4578773/21468923/92ded51e-ca2e-11e6-9f8d-3a551d8998ba.png)

WDYT?

maidmaid committed:
Add Cookie help

maidmaid committed:
Fix cs

wouterj commented on src/Symfony/Bundle/WebProfilerBundle/DependencyInjection/Configuration.php:
```diff
@@ -46,6 +46,7 @@ public function getConfigTreeBuilder()
                 ->end()
                 ->booleanNode('intercept_redirects')->defaultFalse()->end()
                 ->scalarNode('excluded_ajax_paths')->defaultValue('^/(app(_[\\w]+)?\\.php/)?_wdt')->end()
+                ->booleanNode('help')->defaultFalse()->end()
```
As Symfony already has lots of config options, it's often tried to avoid adding many more options. I think this is a case where a config option is not required. Always enabling help will only add some minor question mark icons, which aren't that distracting (and maybe even helpfull for everyone). Always enabling it will also make code a lot easier (avoiding many if statements).

wouterj commented on src/Symfony/Bundle/WebProfilerBundle/Twig/WebProfilerExtension.php:
```diff
@@ -72,6 +76,7 @@ public function getFunctions()
         return array(
             new \Twig_SimpleFunction('profiler_dump', $profilerDump, array('is_safe' => array('html'), 'needs_environment' => true)),
             new \Twig_SimpleFunction('profiler_dump_log', array($this, 'dumpLog'), array('is_safe' => array('html'), 'needs_environment' => true)),
+            new \Twig_SimpleFunction('help', array($this, 'help'), array('is_safe' => array('html'))),
```
I would propose to call this `help_attrs` or the like, to cover the usage of the method more (calling it attrs makes it very clear that this has to be executed inside a tag).

wouterj commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -994,3 +994,60 @@ table.logs .metadata {
+.help {
+    display: inline-block;
+    text-align: center;
+    background-color: #6da581;
```
I propose to make the color a bit lighter (less contrasting with the background, e.g. a light grey). Help icons shouldn't get attention when looking the page, they should only be visible when one needs it.

I'm afraid using the highlight/primary color will make it too distracting.

wouterj reviewed (commented):
I really really like this idea!

(and enjoy your christmas! 🎄)

wouterj commented:
status: needs work

iltar commented:
I'm all for this idea!

ro0NL commented:
:+1: we should keep a single resource though containing all references (think maintenance).

And perhaps we should just provide a list of links (imo. there can be more relevent references then doc+api).

```yaml
# help.yml
cookies:
  - '/some/ref'
  - '/other-ref'

# etc.
```

Also :+1: for making it configurable :)

maidmaid committed:
Add help.yml + remove help config + improve design + refactoring

maidmaid commented:
I have added ``help.yml`` containing all references as @ro0NL suggested, removed ``help`` config, refactored ``help_attr`` twig function and improved design like this :

![screenshot from 2016-12-26 00-32-05](https://cloud.githubusercontent.com/assets/4578773/21473250/c8ec8a5a-cb02-11e6-823c-e65368bf9251.png)

maidmaid committed:
Remove help config

javiereguiluz commented:
I think I like the idea ... but the implementation looks overkill to me. Why can't we just simply put the links in the HTML templates instead of using a YAML config file + Twig functions? (but please, don't make any change yet, this is just a comment)

fabpot commented:
Regarding the implementation: I agree with @javiereguiluz 

Regarding the feature: I'm not sure that people looking at cookies in the web profiler expect to get documentation or API docs from there. Let's see what other @symfony/deciders think about it.

maidmaid commented:
@javiereguiluz the Twig function manages the Symfony version of the project that is included in the documentation URL. These dynamic URL can not be defined directly in templates.
@fabpot Yes, the example of cookies is not the most relevant, but for example links of documentation in the Security section will be very useful.

ro0NL commented:
Should we consider having `Sfjs.set/getPreference` to make this opt-in/out (rather easily)? And perhaps find a clean way to have some simple panel to actually control/expose preferences?

I proposed the YAML file.. and it could indeed be overkill (not sure how many links we expect to end up around templates?). Perhaps a static array in code is better?

fabpot commented:
As you know, we try very hard to not create new option/configuration when not strictly needed. This is a great example of something that should not be configured. If we think it's interesting, it should always be there (and it's not that disturbing anyway). But before going further, I would like to get a list of links we could add and where. I'm still not convinced that this is the best place to add links to the docs.

ro0NL commented:
> This is a great example of something that should not be configured.

Agree. But i think the no. of question icons shown on the same page is crucial here. With the latest design (sorry) i think after a few days the grey dots would start annoy me. So the design should be really slick here.., but that's matter of preference (<punch line here>).

dunglas commented:
I like the idea of making it easy for newcomers to know where to find the relevant documentation.

ogizanagi commented:
@ro0NL : If you fear to be annoyed by these helping dots, I'd rather invite you to use something like [Stylish](https://userstyles.org/) and a custom css rule instead of making it configurable 😄 
(However the dots seem way too big on the current screenshot IMHO 😅)

ro0NL commented:
@ogizanagi eventually yes. But if people start doing that, maybe the design is bad. Note im not saying there's a holy grail or so which works for everbody.

Anyway :+1: for a static feature.

Speaking about;

Im not so sure about API links, they dont seem relevant. And wouldnt it be nicer to have 1 single help icon per page? I.e on top? Containing all references applying to contents for that page, i believe this is the common "approach" in software... making design preferences less an issue.

stof commented:
I agree with @javiereguiluz and @fabpot about this system being overkill.

And I think that links to the doc are more useful than links to the API doc. I would not link to the API doc from the profiler.

sstok commented on src/Symfony/Bundle/WebProfilerBundle/Twig/WebProfilerExtension.php:
```diff
@@ -72,6 +79,7 @@ public function getFunctions()
         return array(
             new \Twig_SimpleFunction('profiler_dump', $profilerDump, array('is_safe' => array('html'), 'needs_environment' => true)),
             new \Twig_SimpleFunction('profiler_dump_log', array($this, 'dumpLog'), array('is_safe' => array('html'), 'needs_environment' => true)),
+            new \Twig_SimpleFunction('help_attrs', array($this, 'help'), array('is_safe' => array('html'))),
```
This should be prefixed with `profiler_` to prevent conflicts with other extensions.

maidmaid commented:
About @fabpot quote:
> But before going further, I would like to get a list of links we could add and where.

Here, *my* list of links proposal:

## Main Title ``Symfony Profiler`` (on top right)

- Title: [Doc](http://symfony.com/doc/current/profiler.html), [Config](http://symfony.com/doc/current/reference/configuration/web_profiler.html#full-default-configuration)

## Panel ``Request / Response``

- Title ``Request / Response`` (missing title): [Component](https://symfony.com/doc/master/components/http_foundation.html)
- Tab ``Request``: [Doc](https://symfony.com/doc/master/components/http_foundation.html#request)
  - Section ``Cookies``: [Doc](https://symfony.com/doc/master/components/http_foundation.html#setting-cookies), [Config](https://symfony.com/doc/current/reference/configuration/framework.html#cookie-lifetime)
  - Section ``Server Parameters``: [PHP](https://secure.php.net/manual/en/reserved.variables.server.php)
- Tab ``Response``: [Doc](https://symfony.com/doc/master/components/http_foundation.html#response)
- Tab ``Session``: [Usage](https://symfony.com/doc/current/controller.html#managing-the-session), [Doc](https://symfony.com/doc/current/session.html), [Config](https://symfony.com/doc/current/reference/configuration/framework.html#session)
- Tab ``Flashes``: [Doc](https://symfony.com/doc/current/controller.html#flash-messages)

## Panel ``Performance``

- Title ``Performance`` (missing title): [Doc](http://symfony.com/doc/current/performance.html)

## Panel ``Logs``

- Title ``Log Messages``: [Doc](http://symfony.com/doc/current/logging.html), [Config](http://symfony.com/doc/current/reference/configuration/monolog.html), [Doc Monolog](https://seldaek.github.io/monolog/#documentation)
- Tab ``Info. & Errors`` and ``Debug``: [Doc](https://github.com/Seldaek/monolog/blob/master/doc/01-usage.md#log-levels)
- Tab ``Silenced PHP Notices``: [Doc](http://symfony.com/doc/current/components/phpunit_bridge.html#trigger-deprecation-notices)

## Panel ``Forms``

- Title ``Form``: [Doc](http://symfony.com/doc/current/forms.html), [Component](http://symfony.com/doc/current/components/form.html), [Config](http://symfony.com/doc/current/reference/configuration/framework.html#reference-framework-form), [Best practices](http://symfony.com/doc/current/best_practices/forms.html)

For each fields:
- Title ``<name field>``: doc of form type (example with ``TextType`` type form: [Doc](http://symfony.com/doc/current/reference/forms/types/text.html))
- Section ``Default Data``: [Doc](http://symfony.com/doc/current/form/events.html#pre-populating-the-form-formevents-pre-set-data-and-formevents-post-set-data)
- Section ``Submitted Data``: [Doc](http://symfony.com/doc/current/form/events.html#submitting-a-form-formevents-pre-submit-formevents-submit-and-formevents-post-submit)
- Section ``Passed Options`` and ``Resolved Options``:
  - for each option, doc of option (example with ``label`` option for ``TextType`` form type: [Doc](http://symfony.com/doc/current/reference/forms/types/text.html#label))
  - for each constraint option, doc of constraint (example with ``NotBlank`` constraint: [Doc](http://symfony.com/doc/current/reference/constraints/NotBlank.html))

## Panel ``Exception``

- Title ``Exceptions``: [Doc](https://symfony.com/doc/current/controller/error_pages.html)

## Panel ``Events``

- Title ``Event Dispatcher``: [Doc](http://symfony.com/doc/current/event_dispatcher.html), [Component](http://symfony.com/doc/current/components/event_dispatcher.html), [Tools](http://symfony.com/doc/current/event_dispatcher.html#debugging-event-listeners)
- Event ``kernel.request``: [Doc](http://symfony.com/doc/current/reference/events.html#kernel-request)
- Event ``kernel.controller``: [Doc](http://symfony.com/doc/current/reference/events.html#kernel-controller)
- Event ``kernel.view``: [Doc](http://symfony.com/doc/current/reference/events.html#kernel-view)
- Event ``kernel.response``: [Doc](http://symfony.com/doc/current/reference/events.html#kernel-response)
- Event ``kernel.finish_request``: [Doc](http://symfony.com/doc/current/reference/events.html#kernel-finish-request)
- Event ``kernel.terminate``: [Doc](http://symfony.com/doc/current/reference/events.html#kernel-terminate)
- Event ``kernel.exception``: [Doc](http://symfony.com/doc/current/reference/events.html#kernel-exception)
- Event ``console.command``: [Doc](http://symfony.com/doc/current/components/console/events.html#the-consoleevents-command-event)
- Event ``console.exception``: [Doc](http://symfony.com/doc/current/components/console/events.html#the-consoleevents-exception-event)
- Event ``console.terminate``: [Doc](http://symfony.com/doc/current/components/console/events.html#the-consoleevents-terminate-event)

## Panel ``Routing``

- Title ``Routing``: [Doc](https://symfony.com/doc/current/routing.html), [Component](https://symfony.com/doc/current/components/routing.html), [Tools](http://symfony.com/doc/current/routing/debug.html), [Best practices](http://symfony.com/doc/current/best_practices/controllers.html#routing-configuration)
- Section ``Route Parameters``: [Doc](https://symfony.com/doc/current/routing.html#adding-wildcard-requirements)

## Panel ``Translation``

- Title ``Translation`` (missing title): [Doc](http://symfony.com/doc/current/translation.html), [Component](http://symfony.com/doc/current/components/translation.html), [Config](http://symfony.com/doc/current/reference/configuration/framework.html#translator), [Tools](http://symfony.com/doc/current/translation/debug.html), [Best practices](http://symfony.com/doc/current/best_practices/i18n.html)
- Box ``Fallback messages``: [Doc](http://symfony.com/doc/current/translation.html#fallback-translation-locales)

## Panel ``Security``

- Title ``Security`` (missing title): [Doc](https://symfony.com/doc/current/security.html), [Component](https://symfony.com/doc/current/components/security.html), [Config](http://symfony.com/doc/current/reference/configuration/security.html), [Best practices](http://symfony.com/doc/current/best_practices/security.html)

Section Security Token:

- Box ``Authenticated``: [Doc](http://symfony.com/doc/current/security.html#initial-security-yml-setup-authentication)
- Property ``Roles``: [Doc](http://symfony.com/doc/current/security.html#roles)
- Property ``Inherited Roles``: [Doc](http://symfony.com/doc/current/security.html#hierarchical-roles)

Section Security Firewall:

- Section ``Security Firewall``: [Doc](http://symfony.com/doc/current/security.html#initial-security-yml-setup-authentication)
- Key ``provider``: [Doc](http://symfony.com/doc/current/security.html#b-configuring-how-users-are-loaded)
- Key ``entry_point``: [Doc](http://symfony.com/doc/current/components/security/firewall.html#entry-points)
- Key ``user_checker``: [Doc](http://symfony.com/doc/current/security/user_checkers.html)
- Key ``access_denied_handler``: [Doc](https://symfony.com/doc/current/security/access_denied_handler.html)
- Key ``listeners``: [Doc](https://symfony.com/doc/current/components/security/firewall.html#firewall-listeners)

Section Security Voters:

- Section ``Security Voters``: [Doc](http://symfony.com/doc/current/security/voters.html)
- Box ``Strategy``: [Doc](http://symfony.com/doc/current/security/voters.html#changing-the-access-decision-strategy)
- Voter class ``Symfony\Component\Security\Core\Authorization\Voter\RoleVoter``: [Doc](https://symfony.com/doc/current/components/security/authorization.html#rolevoter)
- Voter class ``Symfony\Component\Security\Core\Authorization\Voter\AuthenticatedVoter``: [Doc](https://symfony.com/doc/current/components/security/authorization.html#authenticatedvoter)

## Panel ``Twig``

- Title ``Twig`` (missing title): [Doc](http://symfony.com/doc/current/templating.html), [Config](http://symfony.com/doc/current/reference/configuration/twig.html), [Best practices](http://symfony.com/doc/current/best_practices/templates.html), [Doc Twig](http://twig.sensiolabs.org/documentation)

## Panel ``Doctrine``

- Title ``Doctrine`` (missing title): [Doc](http://symfony.com/doc/current/doctrine.html), [Config](http://symfony.com/doc/current/reference/configuration/doctrine.html), [Tools](http://symfony.com/doc/current/doctrine/console.html), [Doc Doctrine](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/)
- Section ``Queries``: [Doc](http://symfony.com/doc/current/doctrine.html#querying-for-objects)
- Sections ``Database Connections`` and ``Entity Managers``: [Doc](http://symfony.com/doc/current/doctrine/multiple_entity_managers.html)
- Section ``Second Level Cache``: [Doc](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/second-level-cache.html)
- Section ``Entities Mapping``: [Doc](http://symfony.com/doc/current/doctrine.html#add-mapping-information)

## Panel ``E-mails``

- Title ``E-mails``: [Doc](http://symfony.com/doc/current/email.html), [Config](http://symfony.com/doc/current/reference/configuration/swiftmailer.html), [Doc Swiftmailer](http://swiftmailer.org/docs/introduction.html)

## Panel ``Debug``

- Title ``Dumped Contents``: [Doc](http://symfony.com/doc/current/templating/debug.html), [Component](http://symfony.com/doc/current/components/var_dumper.html), [Config](http://symfony.com/doc/current/reference/configuration/debug.html)

## Panel ``Configuration``

- Box ``Environment``: [Doc](https://symfony.com/doc/current/configuration/environments.html)
- Box ``Application name``: [Doc](http://symfony.com/doc/current/reference/configuration/kernel.html#kernel-name)
- Section ``Enabled Bundles``: [Doc](http://symfony.com/doc/current/bundles.html)

tjaari commented:
Nice addition 👍

fabpot commented:
I still think this is not the right place for such links. But that's just my opinion of course. @symfony/deciders WDYT?

fabpot commented:
By the way, #21081 seems like a much better idea.

fabpot commented:
Closing as I haven't changed my mind. That's not the right place.
